### PR TITLE
fix(testing): handle more complex projects for react component testing

### LINKED ISSUE #11372: Cypress component testing - multiple issues
## Current Behavior

The following currently seems to not be supported in the component testing configuration preset when running `nx g @nrwl/react:cypress-component-configuration`:
* React SVGR
* Tailwind / PostCSS
* `@testing-library/cypress`

## Expected Behavior
The issues above are supported OOTB in React apps/libs, so they should probably be supported here as well.

## Steps to Reproduce

* Create a workspace @ v14.5.0 with a React library
* Add tailwind & PostCSS configuration
* Add a `styles.css` file that imports tailwind stuff (base, components & utils)
* Run `nx g @nrwl/react:cypress-component-configuration` with the relevant lib
* Add `import @testing-library/cypress/add-commands` to `component.ts`
* Add `import '../../src/styles.css` to `component.ts`
* Run `nx component-test --watch`

Result:
* With testing library, you should see the error `exports is not defined`
* With the CSS file, you should see errors related to the tailwind usage

Here's a repo with the reproduction - https://github.com/danr-za/cypress-ct-nx

### Environment

```sh
  Node : 16.13.0
   OS   : darwin x64
   yarn : 3.2.2
   
   nx : 14.5.0
   @nrwl/angular : Not Found
   @nrwl/cypress : 14.5.0
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.5.0
   @nrwl/eslint-plugin-nx : 14.5.0
   @nrwl/express : Not Found
   @nrwl/jest : 14.5.0
   @nrwl/js : 14.5.0
   @nrwl/linter : 14.5.0
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : 14.5.0
   @nrwl/nx-cloud : 14.2.0
   @nrwl/nx-plugin : Not Found
   @nrwl/react : 14.5.0
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.5.0
   @nrwl/web : 14.5.0
   @nrwl/workspace : 14.5.0
   typescript : 4.7.4
   ---------------------------------------
   Community plugins:
         nx-stylelint: 13.5.2
```

### LINKED ISSUE #11509: @nrwl/react/plugins/component-testing is lacking stuff from react/plugins/webpack
## Current Behavior
`@nrwl/react/plugins/component-testing` is lacking webpack config parts that exists in `@nrwl/react/plugins/webpack` 
Which leads to that code that builds and runs using regular react webpack setup does not build with it out of the box.

Mostly the svgr loader is missing, hot reload, and  maybe more.

Also there is something fishy with the imports, see the workaround to understand what's fishy about them.

## Expected Behavior
Code that works with`@nrwl/react/plugins/webpack` should just work with the `@nrwl/react/plugins/component-testing` preset!

My workaround `cypress.config.ts`:
```ts
import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing';
import * as getWebpackConfigTypeIsBroken from '@nrwl/react/plugins/webpack';
import { defineConfig } from 'cypress';

import type { Configuration } from 'webpack';

const component = nxComponentTestingPreset(__dirname);

fixPartialWebpackConfig(component.devServer.webpackConfig);

export default defineConfig({
  component,
});

/**
 The webpack config in @nrwl/react/plugins/component-testing is incomplete
 Until it will be fixed, we manually add augment it using @nrwl/react/plugins/webpack
 But unfortunately there is a miss-match in the es transpile module configurations and importing it in the normal way is broken,
 so we have a workaround also for that
*/
function fixPartialWebpackConfig(webpackConfig: Configuration) {
  // The exports here are wacked https://github.com/nrwl/nx/blob/12f0f19/packages/react/plugins/webpack.ts
  // mix of ES export + module.exports
  const getWebpackConfig: typeof getWebpackConfigTypeIsBroken.getWebpackConfig =
    // @ts-expect-error see above comments
    getWebpackConfigTypeIsBroken.default;

  const svgToString = /\.svg$/.toString();

  const doWeHaveSvgLoader = webpackConfig.module?.rules?.some((rule) => {
    if (typeof rule === 'string') {
      return false;
    } else {
      return rule.test?.toString() === svgToString;
    }
  });

  /* Kill switch, maybe this is not longer needed */
  if (doWeHaveSvgLoader) {
    throw new Error('---fixPartialWebpackConfig is may no longer needed---');
  }

  // this function actually mutates the config in-place
  getWebpackConfig(webpackConfig);
}

```

Worth noting that my workaround does not make fast refresh to work and maybe more issues.

### PR DESCRIPTION
**Pull request by barbados-clemens**
This change will now use `@nrwl/web:webpack` utilities and requires a build target to be provided in the `@nrwl/cypress:cypress` executor options. Using `nxComponentTestingPreset` will load the `@nrwl/web:webpack` executor options from the provided build target in order to create the webpack configuration for cypress dev server. As of right now, if it's unable to build the webpack configuration, it will fallback to the default configuration that was there before. Which will handle simple cases/can be extended as needed.

When generating a new component testing configuration, a new option for --build-target is available. this allows for manually specifying the target to use for component testing. If one isn't provided the project graph is used to try to find an existing valid build configuration by looking for parent projects that use `@nrwl/web:webpack` (i.e. a React App)

## Potential issues
if working in a workspace without a project using `@nrwl/web:webpack` i.e. a publishable react lib without apps. 2 options.
1. Allow for `@nrwl/cypress:cypress` to accept a full configuration for `@nrwl/web:webpack` / make a new component testing executor that has these options (might not be ideal as new frameworks might have different needs i.e. angular)
2. generate a target that has uses `@nrwl/web:webpack` and pass to the component testing options. i.e. make a dummy app or add a custom `ct-build` target with `@nrwl/web:webpack` in the current project.

## Current Behavior
the default webpack configuration used for cypress component testing (in react) fails when having more "complex" set ups i.e. tailwinds via setup-tailwinds generator.

## Expected Behavior
things automagically work.

## Related Issue(s)

Fixes #11372 
Fixes #11509

### CODE REVIEW

File: packages/react/plugins/component-testing/index.ts [hunk 1]
```diff
@@ -18,162 +40,157 @@ import type { Configuration } from 'webpack';
...
-            react: {
-              runtime: 'automatic',
-            },
-          },
-          loose: true,
-        },
-      },
-    });
-  }
+  options.compiler = options.compiler || 'babel';
```
Comment (barbados-clemens) [inline packages/react/plugins/component-testing/index.ts:108]: use `||=`

Comment (barbados-clemens) [reply packages/react/plugins/component-testing/index.ts:108]: or `??=`

File: packages/react/plugins/component-testing/index.ts [hunk 2]
```diff
@@ -18,162 +40,157 @@ import type { Configuration } from 'webpack';
...
+ */
+function withSchemaDefaults(
+  target: TargetConfiguration<WebWebpackExecutorOptions>
+) {
+  const options = target.defaultConfiguration
+    ? {
+        ...target.options,
+        ...target.configurations[target.defaultConfiguration],
+      }
+    : target.options;
```
Comment (barbados-clemens) [inline packages/react/plugins/component-testing/index.ts:106]: readTargetOptions() will need to build an executor context :(

File: packages/react/plugins/component-testing/index.ts [hunk 3]
```diff
@@ -18,162 +40,157 @@ import type { Configuration } from 'webpack';
...
-  ],
-};
+function getConfigByPath(
+  graph: ProjectGraph,
+  configPath: string
+): ProjectConfiguration {
+  const configFileFromWorkspaceRoot = relative(
+    workspaceRoot,
+    configPath
+  ).replace(extname(configPath), '');
```
Comment (barbados-clemens) [inline packages/react/plugins/component-testing/index.ts:177]: what if dir has `.` in it?

File: packages/react/plugins/component-testing/index.ts [hunk 4]
```diff
@@ -18,162 +40,157 @@ import type { Configuration } from 'webpack';
...
+export function nxComponentTestingPreset(
+  pathToConfig: string,
+  options?: ReactComponentTestingOptions
+) {
+  let webpackConfig;
+  try {
+    const graph = readCachedProjectGraph();
+    const { targets, name } = getConfigByPath(graph, pathToConfig);
+    const targetName = options?.ctTargetName || 'component-test';
+    const mergedOptions = targets[targetName].defaultConfiguration
```
[the remark this answers is not in the document]
Comment (barbados-clemens) [reply packages/react/plugins/component-testing/index.ts:55]: look at readTargetOptions

File: packages/react/src/generators/cypress-component-configuration/cypress-component-configuration.ts [hunk 5]
```diff
@@ -32,6 +19,7 @@ export async function cypressComponentConfigGenerator(
     skipFormat: true,
   });
 
+  await updateProjectConfig(tree, options);
   addFiles(tree, projectConfig, options);
   updateTsConfig(tree, projectConfig);
```
Comment (barbados-clemens) [inline packages/react/src/generators/cypress-component-configuration/cypress-component-configuration.ts:24]: add to cacheable operations

File: packages/react/src/generators/cypress-component-configuration/schema.json [hunk 6]
```diff
@@ -24,6 +24,11 @@
       },
       "x-prompt": "What project should we add Cypress component testing to?"
     },
+    "buildTarget": {
+      "type": "string",
+      "description": "A build target used to configure Cypress component testing in the format of `project:target[:configuration]`. The build target should be from a React app. If not provided we will try to infer it from your projects usage.",
+      "pattern": "^[^:\\s]+:[^:\\s]+(:\\S+)?$"
```
Comment (barbados-clemens) [inline packages/react/src/generators/cypress-component-configuration/schema.json:30]: idea for console? be able to pick a target?

File: packages/cypress/src/utils/find-target-options.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  const graph = await createProjectGraphAsync();
+  const parents = findParentsOfProject(graph, options.project);
+  if (parents.length > 0) {
+    for (const parent of parents) {
+      const [maybeBuildTarget, maybeTargetConfig] = findTargetOptionsInProject(
+        tree,
+        parent.source,
+        options.validExecutorNames
+      );
+      if (maybeBuildTarget && maybeTargetConfig) {
```
[the remark this answers is not in the document]
Comment (barbados-clemens) [reply packages/cypress/src/utils/find-target-options.ts:87]: decided to log if multiple projects found and mention using --build-target can be used to specify the desired target

File: packages/react/src/generators/cypress-component-configuration/cypress-component-configuration.ts (hunk 5 above)
Comment (barbados-clemens) [reply packages/react/src/generators/cypress-component-configuration/cypress-component-configuration.ts:24]: it's in the base generator from @nrwl/cypress 🤦

File: packages/react/src/generators/cypress-component-configuration/schema.json (hunk 6 above)
Comment (barbados-clemens) [reply packages/react/src/generators/cypress-component-configuration/schema.json:30]: jon is already on it 🔥

File: packages/cypress/src/utils/find-target-options.ts [hunk 8]
```diff
@@ -64,65 +65,79 @@ function findInTarget(tree: Tree, options: FindTargetOptions) {
   if (!foundConfig) {
     return projectConfig?.targets?.[target];
   }
-  projectConfig.targets[target].options = {
+
+  const mergedOptions = {
     ...(projectConfig.targets[target]?.options || {}),
     ...projectConfig.targets[target]?.configurations[foundConfig],
   };
```
Comment (AgentEnder) [inline packages/cypress/src/utils/find-target-options.ts:72]: Can this use readTargetOptions? e.g. `readTargetOptions({project, target configuration})`

Comment (AgentEnder) [reply packages/cypress/src/utils/find-target-options.ts:72]: If not, you can remove the `|| {}`

File: packages/cypress/src/utils/find-target-options.ts [hunk 9]
```diff
@@ -64,65 +65,79 @@ function findInTarget(tree: Tree, options: FindTargetOptions) {
...
   for (const targetName in projectConfig.targets) {
     const targetConfig = projectConfig.targets[targetName];
     if (includes.has(targetConfig.executor)) {
-      targetConfig.options = targetConfig.defaultConfiguration
+      const mergedOptions = targetConfig.defaultConfiguration
         ? {
             ...targetConfig.options,
             ...targetConfig.configurations[targetConfig.defaultConfiguration],
           }
         : targetConfig.options;
```
Comment (AgentEnder) [inline packages/cypress/src/utils/find-target-options.ts:132]: `readTargetOptions({project: projectName, target: targetName})`

File: packages/react/plugins/component-testing/index.ts [hunk 10]
```diff
@@ -158,39 +185,68 @@ function buildTargetWebpack(
...
+  targets: Record<string, TargetConfiguration>,
+  projectName: string,
+  targetName: string,
+  configurationName: string
+) {
+  const projectConfigs: { [projectNames: string]: ProjectConfiguration } =
+    Object.entries(graph.nodes).reduce((acc, [key, value]) => {
+      acc[key] = value.data;
+      return acc;
+    }, {});
```
Comment (AgentEnder) [inline packages/react/plugins/component-testing/index.ts:236]: Use `readProjectsConfigurationFromProjectGraph` in `packages/nx/src/project-graph/project-graph.ts`

**Review by AgentEnder — CHANGES_REQUESTED** [review]

**Review by AgentEnder — APPROVED** [review]

**Pull request merged**